Source/WebCore/html/HTMLLabelElement.cpp

```diff
@@ -73,18 +73,42 @@ Ref<HTMLLabelElement> HTMLLabelElement::create(Document& document)
 
 RefPtr<HTMLElement> HTMLLabelElement::control() const
 {
-    auto& controlId = attributeWithoutSynchronization(forAttr);
-    if (controlId.isNull()) {
+    if (!hasAttributeWithoutSynchronization(forAttr)) {
         // Search the children and descendants of the label element for a form element.
         // per http://dev.w3.org/html5/spec/Overview.html#the-label-element
         // the form element must be "labelable form-associated element".
-        for (const auto& labelableElement : descendantsOfType<HTMLElement>(*this)) {
-            if (labelableElement.isLabelable())
-                return const_cast<HTMLElement*>(&labelableElement);
+        for (const HTMLElement& descendant : descendantsOfType<HTMLElement>(*this)) {
+            if (document().settings().shadowRootReferenceTargetEnabled()) {
+                RefPtr<const Element> referenceTarget = descendant.deepShadowRootReferenceTargetOrSelf();
+                if (!referenceTarget || !is<HTMLElement>(referenceTarget))
+                    continue;
+                const HTMLElement* htmlReferenceTarget = dynamicDowncast<HTMLElement>(referenceTarget.get());
+                if (htmlReferenceTarget->isLabelable())
+                    return const_cast<HTMLElement*>(htmlReferenceTarget);
```

Comment:
Another option for this that is a bit shorter and avoids an unnecessary `is<T>` check would be:

```cpp
RefPtr referenceTarget = dynamicDowncast<HTMLElement>(descendant.deepShadowRootReferenceTargetOrSelf());
return referenceTarget && referenceTarget->isLabelable() ? const_cast<HTMLElement*>(referenceTarget.get()) : nullptr;
```